feat: 공연 조회 및 예약 기능 추가

DonggeonMoon opened the pull request:
추가된 기능

* 공연 목록 조회
* 공연 조회
* 예약

DonggeonMoon committed:
Feat/performance (#6)

* chore: port, context-path 설정 추가

* chore: docker-compose yml 파일 추가

* feat: BaseEntity 생성

* feat: EnableJpaAuditing 설정

* feat: Performance entity, repository 추가

* feat: Place entity, repository 추가

* feat: Reservation entity, repository 추가

* feat: Seat entity, repository 추가

* feat: application.yml 작성

도커에서 실행 시 - application.yml
로컬에서 실행 시 - application-local.yml

* feat: bootjar 출력 경로 변경

* fix: reservation 테이블 아이디 embeddedId로 변경

* fix: 도커 컴포즈 파일 수정

* fix: 일반 프로파일 db url 변경

* fix: jdk 주석

* fix: db 테이블 구조 변경

* Create README.md

(cherry picked from commit e805b94f9a5f25467a0d358614f7c40c24382d17)

* feat: 공연 기능 작성

* fix: 공연 응답 변경

* feat: 예매 기능 작성

* feat: 예매 기능 작성

* feat: 공연 기능 작성

* feat: 공연 기능 작성

* feat: 공연 기능 작성

* fix: 카프카 오류 해결

* fix: 리턴 타입 변경

* fix: 예외 enum으로 교체

* fix: dto 접근 제한자 수정

* feat: 예매 기능 작성

---------

Co-authored-by: MALLLAG <87420630+MALLLAG@users.noreply.github.com>

MALLLAG commented on src/main/kotlin/com/performance/domain/performance/dto/PerformanceResponse.kt:
```diff
@@ -0,0 +1,36 @@
+    val availableSeat: Int,
+    val place: PlaceResponse
+) {
+    constructor(performance: Performance) : this(
+        performance.id,
+        performance.name,
+        performance.description,
+        performance.price,
+        performance.startAt,
+        performance.endAt,
+        performance.reservationStartAt,
+        performance.reservationEndAt,
+        performance.canReserve
+                && performance.reservationEndAt.isBefore(LocalDateTime.now())
+                && performance.reservationStartAt.isAfter(LocalDateTime.now()),
+        performance.availableSeat,
+        PlaceResponse(performance.place)
+    )
```
정적 팩토리 함수로 바꾸면 더 좋을듯요

MALLLAG commented on src/main/kotlin/com/performance/domain/performance/service/PerformanceService.kt:
```diff
@@ -0,0 +1,28 @@
+class PerformanceService(
+    private val performanceRepository: PerformanceRepository
+) {
+    @Transactional
```
```suggestion
    @Transactional(readOnly = true)
```

조회는 readOnly 붙여주세요

DonggeonMoon replied:
옙

MALLLAG commented on src/main/kotlin/com/performance/domain/performance/service/PerformanceService.kt:
```diff
@@ -0,0 +1,28 @@
+                PerformanceResponse(it)
+            }
+
+    @Transactional
```
```suggestion
    @Transactional(readOnly = true)
```

여기도 readOnly

MALLLAG commented on src/main/kotlin/com/performance/domain/place/dto/PlaceResponse.kt:
```diff
@@ -0,0 +1,21 @@
+    var totalSeat: Int,
+    var seats: List<SeatResponse>?
+) {
+    constructor(place: Place) : this(
+        place.id,
+        place.name,
+        place.totalSeat,
+        place.seats
+            ?.map {
+                SeatResponse(it)
+            }
+    )
```
여기도 정적 팩토리 함수로

MALLLAG commented on src/main/kotlin/com/performance/domain/reservation/aspect/ReservationAspect.kt:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.stereotype.Component
+import java.time.LocalDateTime
+
+@Aspect
+@Component
+class ReservationAspect(
+    private val performanceRepository: PerformanceRepository,
+    private val reservationRepository: ReservationRepository
+) {
+    @Around("execution(* com..ReservationService.saveReservation(..)) && args(request, ..)")
+    fun checkReservationAvailability(joinPoint: ProceedingJoinPoint, request: ReservationRequest): Any =
+        joinPoint
+            .takeIf {
+                val performance =
+                    performanceRepository.findByIdOrNull(request.performanceId) ?: throw IllegalStateException()
+                performance.canReserve
+                        && performance.reservationEndAt.isAfter(LocalDateTime.now())
+                        && performance.reservationStartAt.isBefore(LocalDateTime.now())
+            }.takeIf {
+                reservationRepository.findByPerformanceIdAndSeatId(
+                    request.performanceId,
+                    request.seatId
+                ).isEmpty()
+            }?.proceed()
+            ?: throw IllegalStateExceptions.RESERVATION_UNAVAILABLE.exception()
+}
```
saveReservation 함수 실행 전후로 검사하는 로직을 aop로 하신 것 같네요.

근데 이 로직은 여러 함수에서 공통적으로 사용하는 함수가 아니라서, 굳이 aop로 할 필요 없을 것 같네요.
그냥 private 함수로 구현하고, 추후에 공통적으로 사용될 때 aop나 valid layer로 분리하는게 좋을듯요.

DonggeonMoon replied:
제약 조건 검사를 비즈니스 로직에서 분리해야 한다고 생각해서 aop를 분리해보았는데 비즈니스 로직에 넣으면 로직이 좀 복잡해지지 않을까요?

MALLLAG replied:
1. 단순한 입력값 유효성 검사가 아닌 persistence layer를 이용한 유효성 검사는 비즈니스 로직과 같이 있는게 더 좋다고 생각합니다. (가독성 면에서도 그렇고, 저는 이것도 비즈니스 로직의 일부라고 생각)
2. 이 로직은 현재 공통적으로 사용되지 않는 로직이고, 미래에도 그럴 일 없다고 생각됩니다. (그래서 저는 오버 엔지니어링이라고 생각)

MALLLAG commented on src/main/kotlin/com/performance/domain/reservation/controller/ReservationController.kt:
```diff
@@ -0,0 +1,26 @@
+        reservationService.saveReservation(request)
+            .let {
+                ResponseEntity.created(
+                    URI.create("/reservations/" + it.id)
```
이건 무슨 의도인가요 ?

DonggeonMoon replied:
201 상태코드를 반환하면서 생성된 자원의 URI를 같이 반환해주었습니다.

MALLLAG replied:
Hateoas라는거죠 ??

MALLLAG commented on src/main/kotlin/com/performance/domain/reservation/service/ReservationService.kt:
```diff
@@ -0,0 +1,31 @@
+    private val performanceRepository: PerformanceRepository,
+    private val seatRepository: SeatRepository
+) {
+    @Transactional
+    fun saveReservation(request: ReservationRequest) =
+        reservationRepository.save(
+            Reservation(
+                id = request.id,
+                email = request.email,
+                performance = performanceRepository.findByIdOrNull(request.performanceId)
+                    ?: throw IllegalStateException(),
+                seat = seatRepository.findByIdOrNull(request.seatId)
+                    ?: throw IllegalStateException(),
+                status = request.status
+            )
+        )
+}
```
```suggestion
    @Transactional
    fun saveReservation(request: ReservationRequest): Reservation {
        val performance = performanceRepository.findByIdOrNull(request.performanceId) ?: throw IllegalStateException()
        val seat = seatRepository.findByIdOrNull(request.seatId) ?: throw IllegalStateException()

        return reservationRepository.save(
            Reservation(
                id = request.id,
                email = request.email,
                performance = performance,
                seat = seat,
                status = request.status
            )
        )
    }
```

kotlin으로 개발좀 하다 보니, 변수 선언하는게 가독성 좋더라구요 ㅎ.,,,.

DonggeonMoon replied:
변수 선언을 하는 게 좀 터 코틀린다운 건가요?

MALLLAG replied:
코틀린다운건 모르겠지만, 변수 선언해주는게 가독성도 좋고 로직 짜기에도 편하더라구요.

MALLLAG commented on src/main/kotlin/com/performance/domain/seat/dto/SeatResponse.kt:
```diff
@@ -0,0 +1,15 @@
+    val seatNumber: String,
+    val isAvailable: Boolean
+) {
+    constructor(seat: Seat) : this(
+        id = seat.id,
+        seatNumber = seat.seatNumber,
+        isAvailable = seat.isAvailable
+    )
```
요기도 정적 팩토리 함수로

MALLLAG commented on src/main/kotlin/com/performance/global/exception/ReservationUnavailableException.kt:
```diff
@@ -0,0 +1,3 @@
+package com.performance.global.exception
+
+class ReservationUnavailableException(message: String) : IllegalThreadStateException(message)
```
IllegalThreadStateException ???

DonggeonMoon replied:
IllegalStateException이 잘못 들어같네요ㅠ 수정하겠습니다.

MALLLAG commented on src/test/kotlin/com/performance/domain/performance/service/PerformanceServiceTest.kt:
```diff
@@ -0,0 +1,68 @@
+
+@ExtendWith(MockitoExtension::class)
+class PerformanceServiceTest {
+    private val performance: Performance = Performance(
+        id = 1,
+        place = Place(
+            id = 1,
+            name = "공연장1",
+            totalSeat = 100,
+            seats = listOf()
+        ),
+        name = "공연1",
+        description = "설명1",
+        price = 100000.toBigDecimal(),
+        startAt = LocalDateTime.of(2023, 12, 25, 19, 0, 0),
+        endAt = LocalDateTime.of(2023, 12, 25, 22, 0, 0),
+        reservationStartAt = LocalDateTime.of(2023, 12, 1, 0, 0, 0, 0),
+        reservationEndAt = LocalDateTime.of(2023, 12, 8, 0, 0, 0, 0),
+        true,
+        100
+    )
```
factory 클래스 만들어서 stub 객체 생성하면 훨씬 더 깔끔합니다.

DonggeonMoon replied:
한 번 찾아보겠습니다. 조언 감사합니다. 😀

MALLLAG commented on src/test/kotlin/com/performance/domain/reservation/service/ReservationServiceTest.kt:
```diff
@@ -0,0 +1,123 @@
+                id = 1,
+                name = "A1",
+                totalSeat = 1,
+                seats = null
+            ),
+            seatNumber = "A1",
+            isAvailable = true
+        ),
+        status = ReservationStatus.PENDING
+    )
+    val performance = Performance(
+        1,
+        place = Place(
+            id = 1,
+            name = "A1",
+            totalSeat = 1,
+            seats = null
+        ),
+        name = "공연1",
+        description = "공연1입니다.",
+        price = 100000.toBigDecimal(),
+        startAt = LocalDateTime.of(2023, 12, 31, 19, 0, 0),
+        endAt = LocalDateTime.of(2023, 12, 31, 21, 0, 0),
+        availableSeat = 100,
+        canReserve = true,
+        reservationStartAt = LocalDate.of(2023, 12, 1).atStartOfDay(),
+        reservationEndAt = LocalDate.of(2023, 12, 7).atTime(23, 59, 59),
+    )
+
+    val seat = Seat(
+        id = 1,
+        place = Place(
+            id = 1,
+            name = "A1",
+            totalSeat = 1,
+            seats = null
+        ),
+        "A1",
+        true
+    )
```
factory 클래스 만들어서 stub 객체 생성하면 훨씬 더 깔끔합니다. 22

MALLLAG commented on src/test/kotlin/com/performance/domain/performance/service/PerformanceServiceTest.kt:
```diff
@@ -0,0 +1,68 @@
+import java.time.LocalDateTime
+import java.util.*
+
+@ExtendWith(MockitoExtension::class)
```
통합테스트로 안하고 단위 테스트로 하신 이유가 있나요

DonggeonMoon replied:
단위테스트를 위주로 테스트로 하고 통합테스트를 나중에 하려고 했는데 일단 단위테스트만 만들어 놓았습니다.

MALLLAG reviewed (commented):
다음부턴 feat branch 파서 작업해주세요 !

DonggeonMoon committed:
feat: 코드 리뷰 반영 (#8)

* chore: port, context-path 설정 추가

* chore: docker-compose yml 파일 추가

* feat: BaseEntity 생성

* feat: EnableJpaAuditing 설정

* feat: Performance entity, repository 추가

* feat: Place entity, repository 추가

* feat: Reservation entity, repository 추가

* feat: Seat entity, repository 추가

* feat: application.yml 작성

도커에서 실행 시 - application.yml
로컬에서 실행 시 - application-local.yml

* feat: bootjar 출력 경로 변경

* fix: reservation 테이블 아이디 embeddedId로 변경

* fix: 도커 컴포즈 파일 수정

* fix: 일반 프로파일 db url 변경

* fix: jdk 주석

* fix: db 테이블 구조 변경

* Create README.md

(cherry picked from commit e805b94f9a5f25467a0d358614f7c40c24382d17)

* feat: 공연 기능 작성

* fix: 공연 응답 변경

* feat: 예매 기능 작성

* feat: 예매 기능 작성

* feat: 공연 기능 작성

* feat: 공연 기능 작성

* feat: 공연 기능 작성

* fix: 카프카 오류 해결

* fix: 리턴 타입 변경

* fix: 예외 enum으로 교체

* fix: dto 접근 제한자 수정

* feat: 예매 기능 작성

* fix: 정적 팩토리 함수, 읽기 전용 트랜잭션 사용

* refactor: aop 로직 서비스로 이동

* refactor: 필요 없어진 예외 클래스 삭제

* feat: 예약 수정, 삭제 기능 추가

* feat: 두 번째 코드 리뷰 반영

변수 빼서 선언하기
예약 가능 상태 처리 로직 엔티티로 이동
예약 소프트 딜리트로 전환
간단한 로직은 표현식 쓰지 말고 블록으로 변경
연관관계 설정 제거
이넘 컨버터 패키지를 컨스턴트 패키지로 이동

findbyidornull 널 처리 할것.

* feat: 코드 리뷰 반영 3

커스텀 예외 및 응답 코드 enum 추가
코드 스타일 수정

---------

Co-authored-by: MALLLAG <87420630+MALLLAG@users.noreply.github.com>
